File: support/verify/verify.cs
using Sheeter;

namespace verify
{
    public static class Verify
    {
        public static void Main()
        {
            var sheet = new Sheeter.Sheeter(new FileLoader());

            Assert(sheet.FromData().GetAwaiter().GetResult());
            Check(sheet, 1000);
            Console.WriteLine("verify success");
        }

        private static void Check(Sheeter.Sheeter sheet, int threads)
        {
            var threadList = new List<Thread>();

            for (var t = 0; t < threads; t++)
            {
                threadList.Add(
                    new Thread(() =>
                    {
                        Assert(sheet.VerifyData.TryGetValue(1, out var actual));
                        Assert(actual.Name1 == 1);
                        Assert(actual.Name2 == 10);
                        Assert(actual.Name3 == 11);
                        Assert(actual.Name4 == 12);
                        Assert(actual.Name5 == 13);
                        Assert(actual.Float == 0.01f);
                        Assert(actual.Ratio.ParseAs<Ratio>().Float32() == 0.01f);
                        Assert(actual.Duration.ParseAs<Duration>().Interval() == TimeSpan.Parse("1.02:03:04.005"));

                        Assert(sheet.VerifyData.TryGetValue(2, out actual));
                        Assert(actual.Name1 == 2);
                        Assert(actual.Name2 == 20);
                        Assert(actual.Name3 == 21);
                        Assert(actual.Name4 == 22);
                        Assert(actual.Name5 == 23);
                        Assert(actual.Float == 0.0001f);
                        Assert(actual.Ratio.ParseAs<Ratio>().Float32() == 0.0001f);
                        Assert(actual.Duration.ParseAs<Duration>().Interval() == TimeSpan.Parse("1.02:03:04.005"));

                        Assert(sheet.VerifyData.TryGetValue(3, out actual) == false);

                        Assert(sheet.VerifyData.TryGetValue(4, out actual));
                        Assert(actual.Name1 == 4);
                        Assert(actual.Name2 == 40);
                        Assert(actual.Name3 == 41);
                        Assert(actual.Name4 == 42);
                        Assert(actual.Name5 == 43);

                        Assert(sheet.VerifyData.TryGetValue(5, out actual));
                        Assert(actual.Name1 == 5);
                        Assert(actual.Name2 == 50);
                        Assert(actual.Name3 == 51);
                        Assert(actual.Name4 == 52);
                        Assert(actual.Name5 == 53);

                        Assert(sheet.MergeData.TryGetValue(1, out actual));
                        Assert(actual.Name1 == 1);
                        Assert(actual.Name2 == 10);
                        Assert(actual.Name3 == 11);
                        Assert(actual.Name4 == 12);
                        Assert(actual.Name5 == 13);

                        Assert(sheet.MergeData.TryGetValue(2, out actual));
                        Assert(actual.Name1 == 2);
                        Assert(actual.Name2 == 20);
                        Assert(actual.Name3 == 21);
                        Assert(actual.Name4 == 22);
                        Assert(actual.Name5 == 23);

                        Assert(sheet.MergeData.TryGetValue(3, out actual) == false);

                        Assert(sheet.MergeData.TryGetValue(4, out actual));
                        Assert(actual.Name1 == 4);
                        Assert(actual.Name2 == 40);
                        Assert(actual.Name3 == 41);
                        Assert(actual.Name4 == 42);
                        Assert(actual.Name5 == 43);

                        Assert(sheet.MergeData.TryGetValue(5, out actual));
                        Assert(actual.Name1 == 5);
                        Assert(actual.Name2 == 50);
                        Assert(actual.Name3 == 51);
                        Assert(actual.Name4 == 52);
                        Assert(actual.Name5 == 53);
                    })
                );
            } // for

            foreach (var itor in threadList)
                itor.Start();

            foreach (var itor in threadList)
                itor.Join();
        }

        private static void Assert(bool condition)
        {
            if (condition == false)
                throw new Exception("verify failed");
        }
    }

    public class FileLoader : Loader
    {
        public string Load(FileName filename)
        {
            return File.ReadAllText(Path.Combine("json", filename.File));
        }

        public void Error(string name, string message)
        {
            Console.WriteLine("file load failed: " + name + ": " + message);
        }
    }
}

File: support/verify/verify.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"sync"
	"time"

	"github.com/yinweli/Sheeter/v3/support/verify/codeGo"
)

func main() {
	sheet := sheeter.NewSheeter(newFileLoader())
	assert(sheet.FromData())
	check(sheet, 1000)
	fmt.Println("verify success")
}

func check(sheet *sheeter.Sheeter, threads int) {
	waitGroup := &sync.WaitGroup{}
	waitGroup.Add(threads)

	for t := 0; t < threads; t++ {
		go func() {
			actual := sheet.VerifyData.Get(1)
			assert(actual != nil)
			assert(actual.Name1 == 1)
			assert(actual.Name2 == 10)
			assert(actual.Name3 == 11)
			assert(actual.Name4 == 12)
			assert(actual.Name5 == 13)
			assert(actual.Float == 0.01)

			ratio, err := sheeter.RunParse[sheeter.Ratio](actual.Ratio)
			assert(err == nil)
			assert(ratio.Float32() == 0.01)

			duration, err := sheeter.RunParse[sheeter.Duration](actual.Duration)
			assert(err == nil)
			assert(duration.Interval() == time.Hour*26+time.Minute*3+time.Second*4+time.Millisecond*5)

			actual = sheet.VerifyData.Get(2)
			assert(actual != nil)
			assert(actual.Name1 == 2)
			assert(actual.Name2 == 20)
			assert(actual.Name3 == 21)
			assert(actual.Name4 == 22)
			assert(actual.Name5 == 23)
			assert(actual.Float == 0.0001)

			ratio, err = sheeter.RunParse[sheeter.Ratio](actual.Ratio)
			assert(err == nil)
			assert(ratio.Float32() == 0.0001)

			duration, err = sheeter.RunParse[sheeter.Duration](actual.Duration)
			assert(err == nil)
			assert(duration.Interval() == time.Hour*26+time.Minute*3+time.Second*4+time.Millisecond*5)

			actual = sheet.VerifyData.Get(3)
			assert(actual == nil)

			actual = sheet.VerifyData.Get(4)
			assert(actual != nil)
			assert(actual.Name1 == 4)
			assert(actual.Name2 == 40)
			assert(actual.Name3 == 41)
			assert(actual.Name4 == 42)
			assert(actual.Name5 == 43)

			actual = sheet.VerifyData.Get(5)
			assert(actual != nil)
			assert(actual.Name1 == 5)
			assert(actual.Name2 == 50)
			assert(actual.Name3 == 51)
			assert(actual.Name4 == 52)
			assert(actual.Name5 == 53)

			actual = sheet.MergeData.Get(1)
			assert(actual != nil)
			assert(actual.Name1 == 1)
			assert(actual.Name2 == 10)
			assert(actual.Name3 == 11)
			assert(actual.Name4 == 12)
			assert(actual.Name5 == 13)

			actual = sheet.MergeData.Get(2)
			assert(actual != nil)
			assert(actual.Name1 == 2)
			assert(actual.Name2 == 20)
			assert(actual.Name3 == 21)
			assert(actual.Name4 == 22)
			assert(actual.Name5 == 23)

			actual = sheet.MergeData.Get(3)
			assert(actual == nil)

			actual = sheet.MergeData.Get(4)
			assert(actual != nil)
			assert(actual.Name1 == 4)
			assert(actual.Name2 == 40)
			assert(actual.Name3 == 41)
			assert(actual.Name4 == 42)
			assert(actual.Name5 == 43)

			actual = sheet.MergeData.Get(5)
			assert(actual != nil)
			assert(actual.Name1 == 5)
			assert(actual.Name2 == 50)
			assert(actual.Name3 == 51)
			assert(actual.Name4 == 52)
			assert(actual.Name5 == 53)

			waitGroup.Done()
		}()
	} // for

	waitGroup.Wait()
}

func assert(condition bool) {
	if condition == false {
		panic(fmt.Errorf("verify failed"))
	} // if
}

func newFileLoader() *fileLoader {
	_, root, _, ok := runtime.Caller(0)

	if ok == false {
		panic(fmt.Errorf("root path failed"))
	} // if

	return &fileLoader{
		path: filepath.Join(filepath.Dir(root), "json"),
	}
}

type fileLoader struct {
	path string
}

func (this *fileLoader) Load(filename sheeter.FileName) []byte {
	path := filepath.Join(this.path, filename.File())
	data, err := os.ReadFile(path)

	if err != nil {
		panic(fmt.Errorf("file loader: %w", err))
	}

	return data
}

func (this *fileLoader) Error(name string, err error) {
	panic(fmt.Errorf("file loader: %w", err))
}
